### PR TITLE
New ftx symbols

### PR DESCRIPTION
**Pull request by samgermain**

### CODE REVIEW

File: js/ftx.js [hunk 1]
```diff
@@ -473,35 +481,79 @@ module.exports = class ftx extends Exchange {
         //
         const result = [];
         const markets = this.safeValue (response, 'result', []);
+        this.oldSymbolMappings = {};
```
Comment (kroitor) [inline js/ftx.js:484]: We don't really need the oldSymbolMappings, i would drop that piece from this PR.
The old market ids will still work on the requesting side.

Comment (kroitor) [reply js/ftx.js:484]: I suggest that we just change the symbols and types in this case without looking into oldSymbolMappings.

Comment (samgermain) [reply js/ftx.js:484]: What if people were using symbols that look like 'ETH-PERP'?

Comment (kroitor) [reply js/ftx.js:484]: Tbh, I would just announce a switch, gather some feedback, and then would switch to new symbols from, say, 2022.
However, if you think that we should keep supporting them both in the long run, then we should probably unify the symbols-aliases in the base classes across the lib. Personally, I think just switching is fine, if the users are aware and prepared.

Comment (kroitor) [reply js/ftx.js:484]: On a second thought, we could change safeSymbol/safeMarket to account for a user-controllable property that would either return new symbols or old symbols. That would be cleaner, and we would not need the extra helper methods here and there.

Comment (samgermain) [reply js/ftx.js:484]: The old symbols for ftx are also the exchange ids, so those should work anyway shouldn't they?

Comment (kroitor) [reply js/ftx.js:484]: @samgermain yes, they should be ok.

Comment (samgermain) [reply js/ftx.js]: > @samgermain yes, they should be ok.

Ok, I remvoed `oldSymbolMappings`, still don't know what to do about `Trump2024` and `Bolsonaro2022` other than hard code them. They have expiry's of `Trump2024` and `Bolsonaro2022` right now in CCXT, but actual expiry's of oct 2024, and oct 2022

**Comment by frosty00** [comment]
let's that function and only parse the symbols in `fetchMarkets` like how it is done in every other exchange

**Comment by frosty00** [comment]
in other words - we will not parse the symbols of delisted markets

**Comment by samgermain** [comment]
> let's that function and only parse the symbols in `fetchMarkets` like how it is done in every other exchange

I don't understand what you mean sorry.

> in other words - we will not parse the symbols of delisted markets

How do I know which ones are delisted?

**Comment by frosty00** [comment]
I mean let's only parse the market inside of `fetchMarket`

**Comment by frosty00** [comment]
Can we use future instead of futures?

**Comment by frosty00** [comment]
it doesn't really make sense to deprecate future when imo it is the correct grammatical term and also it is used on other exchanges such as binance, i would rather deprecate futures

**Comment by samgermain** [comment]
> I mean let's only parse the marketId inside of `fetchMarket`

Ok, I wasn't sure what to do about `BOLSONARO2022` and `TRUMP2024`, so I created a mapping in options, this might be the best solution, seeing as no expiry date is returned, but there definately is an expiry date for these contracts

Also the FTX expiry dates don't contain the year, is that something we want to add? They're just `1231` or `2022Q4` or something, we could write some functionality that gets the full expiry date

**Comment by frosty00** [comment]
@samgermain we should not have a hardcode mapping, instead in the fetchMarkets we should parse the expiry date properly and use the `this.yymmdd` function to convert a timestamp into the unified futures date format

**Comment by samgermain** [comment]
> @samgermain we should not have a hardcode mapping, instead in the fetchMarkets we should parse the expiry date properly and use the `this.yymmdd` function to convert a timestamp into the unified futures date format

There's no way to do that with the president tokens, on their website they state that the expiry date is election day, for US president tokens, that's November 5 every 4 years except in rare and unusual circumstances, but for Brazil it looks like the date is variable, so there's no way to make it dynamic if there were to be a `BOLSONARO2026` contract or something

File: js/ftx.js [hunk 2]
```diff
@@ -478,30 +482,70 @@ module.exports = class ftx extends Exchange {
...
+            const swap = (this.safeString (splitName, 1) === 'PERP');
+            const option = false;
+            let futures = false;
+            let type = 'spot';
+            let expiry = undefined;
+            if (swap) {
+                type = 'swap';
+                symbol = base + '/' + quote + ':' + settleId;
+            } else if (contract) {
+                type = 'futures';
```
Comment (frosty00) [inline js/ftx.js:504]: let's not change future -> futures for the sake of it

Comment (samgermain) [reply js/ftx.js:504]: > let's not change future -> futures for the sake of it

It should be unified across exchanges

File: js/ftx.js [hunk 3]
```diff
@@ -478,30 +482,70 @@ module.exports = class ftx extends Exchange {
...
+                'settleId': settleId,
                 'type': type,
-                'future': (type === 'future'),
-                'spot': (type === 'spot'),
-                'active': active,
-                'precision': precision,
+                'spot': spot,
+                'margin': margin,
+                'swap': swap,
+                'future': futures, // Deprecated, please use 'futures'
```
Comment (samgermain) [inline js/ftx.js:531]: I also have this, so that `future` will still work

Comment (frosty00) [reply js/ftx.js:531]: no it won't lets restore this aspect of the api so `type` is future pls

**Comment by frosty00** [comment]
https://docs.ftx.com/#list-all-futures

we should probably also parse the info from here then

**Comment by samgermain** [comment]
> it doesn't really make sense to deprecate future when imo it is the correct grammatical term and also it is used on other exchanges such as binance, i would rather deprecate futures

@kroitor Should we deprecate the term `futures`? If we should I can go through and change `gateio`, `huobi` and other exchanges that have it

**Comment by kroitor** [comment]
> @kroitor Should we deprecate the term futures? If we should I can go through and change gateio, huobi and other exchanges that have it

Yes, I think we should standardize it at last, however, if possible please submit one exchange per PR.

**Comment by samgermain** [comment]
@kroitor Any opinion on the president tokens?

File: js/ftx.js [hunk 4]
```diff
@@ -471,37 +471,122 @@ module.exports = class ftx extends Exchange {
...
+                type = 'swap';
+                symbol = base + '/' + quote + ':' + settleId;
+            } else if (contract) {
+                type = 'futures';
+                // TODO: delete
+                if (!expiry || !settleId || !base || !quote) {
+                    type = 'breakpoint';
+                    const y = 'stop_here';
+                    const x = 'wtf';
+                }
```
Comment (samgermain) [inline js/ftx.js:554]: I can't figure out this error, if I set a breakpoint inside this conditional, the breakpoint is never hit, meaning all 4 variables are defined, but I'm getting

```
Testing { exchanges: ["ftx"], symbol: "all", keys: { '--js': true, '--php': false, '--python': false, '--python-async': true, '--php-async': true, '--python-sync': true, '--php-sync': true }, maxConcurrency: 5 } (run-tests.js:255)
[100%] Testing ftx FAIL (testExchange @ run-tests.js:172)

FAILED ftx (Python 3 Async): (explain @ run-tests.js:190)

	ftx using proxy ``
	<class 'TypeError'> can only concatenate str (not "NoneType") to str 
	
	  test/test_async.py", line 558, in <module>
	    asyncio.run(main())
	
	  python3.10/asyncio/runners.py", line 44, in run
	    return loop.run_until_complete(main)
	
	  python3.10/asyncio/base_events.py", line 641, in run_until_complete
	    return future.result()
	
	  test/test_async.py", line 545, in main
	    await try_all_proxies(exchange, proxies)
	
	  test/test_async.py", line 483, in try_all_proxies
	    await load_exchange(exchange)
	
	  test/test_async.py", line 367, in load_exchange
	    await exchange.load_markets()
	
	  async_support/base/exchange.py", line 199, in load_markets
	    raise e
	
	  async_support/base/exchange.py", line 195, in load_markets
	    result = await self.markets_loading
	
	  async_support/base/exchange.py", line 185, in load_markets_helper
	    markets = await self.fetch_markets(params)
	
	  async_support/ftx.py", line 566, in fetch_markets
	    symbol = base + '/' + quote + ':' + settleId + '-' + expiry (explain @ run-tests.js:193)
FAIL ["ftx"] (run-tests.js:271)
All done, 1 failed (run-tests.js:276)
```

File: js/base/Exchange.js [hunk 5]
```diff
@@ -2085,4 +2086,18 @@ module.exports = class Exchange {
         params = this.omit (params, 'type');
         return [ type, params ];
     }
+
+    yearMonthDay (dateStr) {
```
Comment (kroitor) [inline js/base/Exchange.js:2090]: We don't need this, cause we have yyyymmdd, yymmdd, ymd in the base classes.

Comment (samgermain) [reply js/base/Exchange.js:2090]: `yyyymmdd` returns `2021-12-31` instead of `20211231`

Comment (kroitor) [reply js/base/Exchange.js:2090]: @samgermain it has the secont `infix` argument for that, it's a dash `-` by default, but you can set it to `-`:

- https://github.com/ccxt/ccxt/blob/master/js/base/functions/time.js#L161

File: js/ftx.js (hunk 4 above)
Comment (samgermain) [reply js/ftx.js:554]: @kroitor Any idea about this?

Comment (kroitor) [reply js/ftx.js:554]: One of `base`, `quote`, `settleId` or `expiry` is missing in one of markets?

Comment (samgermain) [reply js/ftx.js:554]: I tried setting a breakpoint inside the conditional but it is never hit

Comment (kroitor) [reply js/ftx.js]: In cases like these I just add `console.log (base, quote, settleId, expiry)`, run the test locally and from there it becomes obvious.

**Comment by kroitor** [comment]
Related: #10931

File: js/ftx.js [hunk 6]
```diff
@@ -475,37 +475,123 @@ module.exports = class ftx extends Exchange {
...
+        //
         const result = [];
         const markets = this.safeValue (response, 'result', []);
+        const allFutures = this.safeValue (allFuturesResponse, 'result', []);
+        const allFuturesDict = {};
+        for (let i = 0; i < allFutures.length; i++) {
+            const futureDict = allFutures[i];
+            const id = this.safeString (futureDict, 'name');
+            allFuturesDict[id] = futureDict;
+        }
```
Comment (kroitor) [inline js/ftx.js:532]: We can do lines 526-532 with a one-liner call to `this.indexBy (allFutures, 'name')`. See how it's used in the other exchanges.

File: js/ftx.js [hunk 7]
```diff
@@ -475,37 +475,123 @@ module.exports = class ftx extends Exchange {
...
             const quoteId = this.safeString (market, 'quoteCurrency', 'USD');
-            const type = this.safeString (market, 'type');
+            const settleId = contract ? 'USD' : undefined;
             const base = this.safeCurrencyCode (baseId);
             const quote = this.safeCurrencyCode (quoteId);
+            const settle = this.safeCurrencyCode (settleId);
+            const spot = !contract;
+            const margin = !contract;
+            const perpetual = this.safeValue (future, 'perpetual');
+            const swap = (perpetual === true);
```
Comment (frosty00) [inline js/ftx.js:548]: is this line necessary?

**Pull request merged**